### DISCUSSION TITLE
关于同步功能的讨论

### DISCUSSION THREAD

**Discussion by mcthesw** [mcthesw/game-save-manager#66] [Ideas]
云端备份、同步是一个非常重要的功能，在社群中呼声比较高，因此开设此讨论帖，希望能完成备份与同步的构想，目前的所有构想基于 #61 这个issue，需求有以下几项

1. 云备份
    1. 兼容WebDav、S3等协议的云端存储
    3. 能够定期上传到云端
2. 云同步（需要首先实现云备份的一切功能）
    1. 对于不同机器，实现位置匹配
    2. 对于双方分别产生的新存档，同步时进行合并

根据以上需求，需要制作这些功能：

- 需求1.1，文件交互部分可以使用OpenDAL实现，需要在设置页面开放一个设置块
- 需求1.2，可以在每次启动时同步，之后可以手动同步或开放由用户主动开启的定时同步
- 需求2.1，参考 @AsterNighT 的说法，两类需要处理的文件分别是绝对路径和相对路径，相对路径中的使用用户目录的情况在添加时可以做一下检测，但是具体如何实现我还不清楚，有待讨论；而依赖于游戏路径的情况可以靠强制要求设置游戏所在位置来保存（但是这样需要用户在添加时就确定好是那种路径），这里需要讨论的是如何在存档管理器中加入这个功能，我目前的想法是在 `SaveUnit` 中加入一个字段描述路径是何种路径，或将 `SaveUnit.path` 更改成二选一的enum
- 需求2.2，可以通过简单的对比二者的json文件，只需要做一个简单的diff就行了，以date为唯一标志（不过这样是不是不太好？）

在实现了以上功能后应该就可以完成云备份和存档这一期待已久的功能了，在这之前需要理清上述问题，各位如果有什么看法欢迎交流

**Comment by mcthesw** [comment]
![aa5013965d5c96cdfcc12ae1d6f1ee6e](https://github.com/mcthesw/game-save-manager/assets/61224072/aaf7e64b-1480-4f42-8f41-c99f8de70701)
前端界面大概制作了一下，但是后端制作的时候遇到了很大的问题，大家有什么看法吗？
遇到云端和本地不同时，该怎么处理？比如云端是ACDE，本地是ABCDF，我觉得最好的办法还是记录下来各端进行了什么操作，然后在合并时根据这些操作来处理；或者因为一个人一次肯定只能在一个电脑上玩，干脆禁止这样产生多分支。
我需要一个比较具体的方案来解决这个问题，有没有现成的方案或者开源项目参考？

**Comment by mcthesw** [comment]
同步方案1：使用transaction机制，不只是同步他们的状态，而是同步他们到达这个状态所用的步骤
同步方案2：本地优先，且云端不自动删除旧存档，每次同步diff一下 `./save_data/游戏名/Backups.json` 就行了
同步方案3：软件启动时就强制同步一次，且执行删除、增加操作后立刻发送到云端，要求执行相同操作

**Comment by AsterNighT** [comment]
能不能直接把冲突的版本都保存下来，让用户自己选择用哪个版本呢？毕竟云端也不一定只有一个存档。

**Reply by mcthesw** [reply]
这个也可以，但是会比较麻烦了

**Reply by mcthesw** [reply]
目前在做的方案是上述的同步方案2+同步方案3

**Comment by shixuev5** [comment]
云备份支持下阿里云这类网盘？

**Reply by mcthesw** [reply]
阿里云等国内常用云盘对WebDAV不提供支持，但是有第三方软件可以绕过这个。
如果本项目强行支持国内这些云盘，会有以下三个问题：

1. 开发难度高（可能需要一定的逆向）
2. 存在法律风险
3. 更新更为被动（他更新可能会导致我也要更新）

因此，我认为这类功能不方便加入该软件，如果坚持使用这类云盘，请使用alist这类能将云盘代理为WebDAV服务的第三方软件，本软件预计只会添加常用标准的云存储协议（类似WebDAV和S3）

**Reply by shixuev5** [reply]
好的，没问题。基于 git 的代码管理模型比较合适。不知道 webdav、s3 是否都需要自己实现。

不同设备： 不同分支（根据不同设备硬件生成唯一标识）
相同设备：本地推送云端（push）、云端同步本地（pull） 自动还是手动取决于用户配置
版本对比：git commit id 是否一致
同步记录：提交记录（git commit）
存档版本：基于提交记录打 tag（git tag）

**Reply by shixuev5** [reply]
自己实现的话，就需要通过中间层来转换，通过目录嵌套来区分。可以参考这个

https://www.bilibili.com/read/cv10351656/

**Reply by mcthesw** [reply]
目前云存储基于OpenDAL这个库，因此支持不同的协议我只需要做比较简单的适配，实际上我也有在考虑将本地操作也使用OpenDAL来实现，这样可以实现一层抽象，让多种不同的存储方式有统一的操作方式，比如可以做到`backends.for_each(|x|x.backup(saves))`类似的操作

有关Git这样的管理模型我觉得对于存档管理来说太庞大了，不过游戏存档确实很符合分支的管理方式，我比较想实现一套基于事务的机制，但是实在知识不足，一直没有办法做出来，你说的“根据不同设备硬件生成唯一标识”、推拉策略、版本对比等我觉得是非常有用的，我一直在想当前存档结构到时候如何改进比较好，不过改动可能需要做比较复杂的版本升级适配，得仔细考虑。

要是有好的想法欢迎交流！要是有相关的资料也欢迎分享

**Comment by Font8036** [comment]
这个功能，能支持自己搭建的HTTP服务吗？比如我自己就在使用开源项目HFS的HTTP服务（喜欢用电脑下片，然后用HTTP在手机上直接看）。

**Reply by mcthesw** [reply]
目前支持S3和WebDav，这两个都可以自行搭建，其他云后端尚不支持，不过如果[OpenDAL](https://opendal.apache.org)支持的话比较好加入
